rule/package_directory_mismatch.go

```diff
@@ -0,0 +1,146 @@
+package rule
+
+import (
+	"fmt"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/mgechev/revive/lint"
+)
+
+// PackageDirectoryMismatchRule detects when package name doesn't match directory name.
+type PackageDirectoryMismatchRule struct {
+	ignoredDirs *regexp.Regexp
+}
+
+const defaultIgnoredDirs = "testdata"
+
+// Configure the rule to exclude certain directories.
+func (r *PackageDirectoryMismatchRule) Configure(arguments lint.Arguments) error {
+	if len(arguments) < 1 {
+		return r.buildIgnoreRegex([]string{defaultIgnoredDirs})
+	}
+
+	args, ok := arguments[0].(map[string]any)
+	if !ok {
+		return fmt.Errorf("invalid argument type: expected map[string]any, got %T", arguments[0])
+	}
+
+	for k, v := range args {
+		if !isRuleOption(k, "ignore-directories") {
+			return fmt.Errorf("unknown argument %s for %s rule", k, r.Name())
+		}
+		ignoredDirs, ok := v.([]string)
+		if !ok {
+			return fmt.Errorf("invalid value %v for argument %s of rule %s, expected []string got %T", v, k, r.Name(), v)
+		}
+		return r.buildIgnoreRegex(ignoredDirs)
+	}
+
+	return nil
+}
+
+func (r *PackageDirectoryMismatchRule) buildIgnoreRegex(ignoredDirs []string) error {
+	if len(ignoredDirs) == 0 {
+		return nil
+	}
+
+	patterns := make([]string, len(ignoredDirs))
+	for i, dir := range ignoredDirs {
+		patterns[i] = regexp.QuoteMeta(dir)
+	}
+	pattern := "(" + strings.Join(patterns, "|") + ")"
+
+	var err error
+	r.ignoredDirs, err = regexp.Compile(pattern)
+	if err != nil {
+		return fmt.Errorf("failed to compile regex for ignored directories: %w", err)
+	}
+
+	return nil
+}
+
+// skipDirs contains directory names that should be unconditionally ignored when checking.
+// These entries handle edge cases where filepath.Base might return these values.
+var skipDirs = map[string]struct{}{
+	".": {}, // Current directory
+	"/": {}, // Root directory
+	"":  {}, // Empty path
+}
+
+// semanticallyEqual checks if package and directory names are semantically equal to each other.
+func (PackageDirectoryMismatchRule) semanticallyEqual(packageName, dirName string) bool {
+	return normalizePath(dirName) == normalizePath(packageName) ||
+		normalizePath(dirName) == normalizePath("go"+packageName)
```

Comment:
Shouldn't be this? 🤔


```suggestion
		normalizePath(dirName) == normalizePath("go-"+packageName)
```

Reply:
> In fact the call to normalizePath is not necessary (unless you do "go-")

It's still necessary to make this case work: `package foo_bar` with `dir go-foobar`. This testcase is missing.

Reply:
But with this https://github.com/mgechev/revive/pull/1429#discussion_r2264628892, it's not needed.